internal/builders/golang/build_test.go

```diff
@@ -0,0 +1,299 @@
+package golang
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"runtime"
+	"testing"
+
+	api "github.com/goreleaser/goreleaser/build"
+	"github.com/goreleaser/goreleaser/config"
+	"github.com/goreleaser/goreleaser/context"
+	"github.com/goreleaser/goreleaser/internal/testlib"
+	"github.com/stretchr/testify/assert"
+)
+
+var runtimeTarget = runtime.GOOS + "_" + runtime.GOARCH
+
+func TestWithDefaults(t *testing.T) {
+	for name, testcase := range map[string]struct {
+		build   config.Build
+		targets []string
+	}{
+		"full": {
+			build: config.Build{
+				Binary: "foo",
+				Goos: []string{
+					"linux",
+					"windows",
+					"darwin",
+				},
+				Goarch: []string{
+					"amd64",
+					"arm",
+				},
+				Goarm: []string{
+					"6",
+				},
+			},
+			targets: []string{
+				"linux_amd64",
+				"darwin_amd64",
+				"windows_amd64",
+				"linux_arm_6",
+			},
+		},
+		"empty": {
+			build: config.Build{
+				Binary: "foo",
+			},
+			targets: []string{
+				"linux_amd64",
+				"linux_386",
+				"darwin_amd64",
+				"darwin_386",
+			},
+		},
+	} {
+		t.Run(name, func(tt *testing.T) {
+			var config = config.Project{
+				Builds: []config.Build{
+					testcase.build,
+				},
+			}
+			var ctx = context.New(config)
+			var build = Default.WithDefaults(ctx.Config.Builds[0])
+			assert.ElementsMatch(t, build.Targets, testcase.targets)
+		})
+	}
+}
+
+func TestDefaultAndBuild(t *testing.T) {
+	folder, back := testlib.Mktmp(t)
+	defer back()
+	writeGoodMain(t, folder)
+	var config = config.Project{
+		Builds: []config.Build{
+			{
+				Binary: "foo",
+				Targets: []string{
+					"linux_amd64",
+					"darwin_amd64",
+					"windows_amd64",
+					"linux_arm_6",
+				},
+			},
+		},
+	}
+	var ctx = context.New(config)
+	var build = ctx.Config.Builds[0]
+	for _, target := range build.Targets {
+		var err = Default.Build(ctx, build, api.Options{
+			Target: target,
+			Name:   build.Binary,
+			Path:   filepath.Join(folder, "dist", target, build.Binary),
+		})
+		assert.NoError(t, err)
+	}
+	assert.Len(t, ctx.Artifacts.List(), len(build.Targets))
```

Comment:
We don't actually check if the artifacts are the ones we expect here, but I guess that's okay for now.